core/source/Number.mint

```diff
@@ -42,17 +42,21 @@ module Number {
   fun fromString (input : String) : Maybe(Number) {
     `
     (() => {
+      if (typeof #{input} !== 'string') {
+        return #{Maybe::Nothing}
+      }
+
       let value = Number(#{input})
 
       if (
         Number.isNaN(value)
         ||
-        Number.isNaN(value - parseFloat(#{input}))  // excludes "Infinity", "-Infinity", ""
+        #{input}.trim() === ''
```

Comment:
This check should be done before constructing `value`.

Reply:
Actually, I tested this and the version with late string trimming runs a bit faster: (check console)

```html
<script>
let x, startTime, endTime


function runParseFloat(times) {
    for (let y=0; y < times; y++) {
        x = parseFloat("1")
        x = parseFloat("1.0")
        x = parseFloat(" 1 ")
        x = parseFloat("1e-2")
        x = parseFloat("1E+2")
        x = parseFloat("1.0 xxx")
        x = parseFloat("")
        x = parseFloat("Infinity")
        x = parseFloat("-Infinity")
        x = parseFloat("x")
    }
}
startTime = performance.now()
runParseFloat(10000000)
endTime = performance.now()
console.log(`Call to runParseFloat took ${endTime - startTime} milliseconds`)



function Number_FromStringEarlyTrim(input) {
    if (typeof input !== 'string' || input.trim() === '') {
        return null
    }

    let value = Number(input)

    if (
        Number.isNaN(value)
    ) {
        return null
    }

    return value
}
function runNumber_FromStringEarlyTrim(times) {
    for (let y=0; y < times; y++) {
        x = Number_FromStringEarlyTrim("1")
        x = Number_FromStringEarlyTrim("1.0")
        x = Number_FromStringEarlyTrim(" 1 ")
        x = Number_FromStringEarlyTrim("1e-2")
        x = Number_FromStringEarlyTrim("1E+2")
        x = Number_FromStringEarlyTrim("1.0 xxx")
        x = Number_FromStringEarlyTrim("")
        x = Number_FromStringEarlyTrim("Infinity")
        x = Number_FromStringEarlyTrim("-Infinity")
        x = Number_FromStringEarlyTrim("x")
        x = Number_FromStringEarlyTrim("1x")
    }
}
startTime = performance.now()
runNumber_FromStringEarlyTrim(10000000)
endTime = performance.now()
console.log(`Call to runNumber_FromStringEarlyTrim took ${endTime - startTime} milliseconds`)




function Number_FromString(input) {
    if (typeof input !== 'string') {
        return null
    }

    let value = Number(input)

    if (
        Number.isNaN(value)
        ||
        input.trim() === ''
    ) {
        return null
    }

    return value
}
function runNumberFromString(times) {
    for (let y=0; y < times; y++) {
        x = Number_FromString("1")
        x = Number_FromString("1.0")
        x = Number_FromString(" 1 ")
        x = Number_FromString("1e-2")
        x = Number_FromString("1E+2")
        x = Number_FromString("1.0 xxx")
        x = Number_FromString("")
        x = Number_FromString("Infinity")
        x = Number_FromString("-Infinity")
        x = Number_FromString("x")
        x = Number_FromString("1x")
    }
}
startTime = performance.now()
runNumberFromString(10000000)
endTime = performance.now()
console.log(`Call to runNumberFromString took ${endTime - startTime} milliseconds`)




function runNumber(times) {
    for (let y=0; y < times; y++) {
        x = Number("1")
        x = Number("1.0")
        x = Number(" 1 ")
        x = Number("1e-2")
        x = Number("1E+2")
        x = Number("1.0 xxx")
        x = Number("")
        x = Number("Infinity")
        x = Number("-Infinity")
        x = Number("x")
    }
}
startTime = performance.now()
runNumber(10000000)
endTime = performance.now()
console.log(`Call to runNumber took ${endTime - startTime} milliseconds`)





// console.log(Number_FromString('Infinity'))
// console.log(Number_FromString(''))
// console.log(Number_FromString('   '))
// console.log(Number_FromString(true))
// console.log(Number_FromString("1"))
// console.log(Number_FromString("1.0"))
// console.log(Number_FromString(" 1 "))
// console.log(Number_FromString("1e-2"))
// console.log(Number_FromString("1E+2"))
// console.log(Number_FromString("1.0 xxx"))
// console.log(Number_FromString(""))
// console.log(Number_FromString("Infinity"))
// console.log(Number_FromString("-Infinity"))
// console.log(Number_FromString("x"))

</script>
```

Reply:
There is actually a bug anywhere in my code which distorts the results, but I can not find it. (some hidden JavaScript optimization?) The following code gives the realistic (and expected) results:

```html
<script>

    let VALUES = ["1", "1.0", " 1 ", "1e-2", "1E+2", "1.0 xxx", "", "Infinity", "-Infinity", "x"],
        TIMES = 10000000

    function run(desc, func) {
        let startTime, endTime, x, y, values = VALUES, times = TIMES
        startTime = performance.now()
        for (y = 0; y < times; y++) {
            for (value of values) {
                x = func(value)
            }
        }
        endTime = performance.now()
        console.log(`${times} calls to ${desc} take ${endTime - startTime} milliseconds`)
    }

    run('parseFloat', parseFloat)

    run('Number', Number)

    run('NumberFromStringLateTrim', function (input) {
        if (typeof input !== 'string') {
            return null
        }
        let value = Number(input)
        if (
            Number.isNaN(value)
            ||
            input.trim() === ''
        ) {
            return null
        }
        return value
    })

    run('NumberFromString', function (input) {
        if (typeof input !== 'string' || input.trim() === '') {
            return null
        }
        let value = Number(input)
        if (
            Number.isNaN(value)
        ) {
            return null
        }
        return value
    })

    run('NumberFromStringNoTypecheck', function (input) {
        let value = Number(input)
        if (
            Number.isNaN(value)
            ||
            input.trim() === ''
        ) {
            return null
        }
        return value
    })

    // run('', function (input) {
    // })

</script>
```

In my browser this results in

```
10000000 calls to parseFloat take 2594.7000000001863 milliseconds
10000000 calls to Number take 2872.5 milliseconds
10000000 calls to NumberFromStringEarlyTrim take 3619.5 milliseconds
10000000 calls to NumberFromString take 3617.199999999255 milliseconds
10000000 calls to NumberFromStringNoTypecheck take 3541.600000000559 milliseconds
10000000 calls to Angular Version take 3245.5999999996275 milliseconds
```

We can see here

* As expected, the native functions run fastest (and `parseFloat` is actually the fastest).
* It is actually not relevant if the trimming happens early or late, so I will change the commit to do it early, like @Sija suggested. It is more readable indeed.
* The type checking is not expensive.
* The Angular Version (original commit) is slightly faster than the other non-native functions

My assumption that the code runs faster now was wrong, it actually runs a bit slower. Considerable, if you let it run ten million times, *looks worried*, haha, no.